packages/flutter_tools/lib/src/android/adb.dart

```diff
@@ -112,6 +118,22 @@ class Adb {
     return controller.stream;
   }
 
+  Future _populateDeviceNames(List<AdbDevice> devices) async {
+    for (AdbDevice device in devices) {
+      if (device.modelID == null) {
+        if (_idToNameCache[device.id] == null)
+          await _populateDeviceCache();
+        device.modelID = _idToNameCache[device.id];
```

Comment:
It's a different `AdbDevice` instance. The adb server notification for new devices just includes the shorthand for device info (`05a02bac	device`). When we ask it to list all devices, we ask for the long-hand version - `host:devices-l`, which returns `05a02bac               device usb:336592896X product:razor model:Nexus_7 device:flo`. That has the actual name of the device.